File: tests/cpp/unit/testUnitInterpolator.cpp
#include <algorithm>
#include <cmath>
#include <gmock/gmock.h>
#include <gtest/gtest.h>
#include "dsp.h"

#define private public
#include "firFilter.h"
#include "interpolator.h"
#undef private

using namespace std;
using ::testing::_;
using ::testing::Return;

// Define the mock object.
class HalfBandMock : public HalfBand {
    public:
        HalfBandMock(int nPoints = 8192) : HalfBand(nPoints) {}
        MOCK_METHOD(vector<double>, Call, (double AdB, double Fpass), ());
        vector<double> operator()(double AdB, double Fpass) override { return Call(AdB, Fpass); }
};

// Define the test object.
class TestInterpolator : public ::testing::Test {
    protected:
        dsp::ComplexFFT fft;
        Interpolator interpolator;
};

TEST_F(TestInterpolator, validOutput) {
    // Define the output.
    double fmax = 400.0;
    double fs = 1000.0;
    vector<complex<double>> input(static_cast<int>(fs), complex<double>(0.0, 0.0));
    for (uint n = 0; n < static_cast<int>(fs); n++) {
        input[n] = complex<double>(sin(2 * M_PI * fmax * static_cast<double>(n) / fs));
    }

    // Mock the HalfBand filter.
    vector<double> halfbandResult = {1.0};
    delete interpolator.halfband;
    interpolator.halfband = new HalfBandMock();
    HalfBandMock * halfbandMock = dynamic_cast<HalfBandMock*>(interpolator.halfband);
    EXPECT_CALL(*halfbandMock, Call(_, _)).Times(4).WillRepeatedly(Return(halfbandResult));

    // Compute the result.
    vector<complex<double>> output = interpolator(120.0, fmax, fs, input);
    output = fft.fft(output);
    uint N = static_cast<uint>(pow(2U, interpolator.getN()) * fs);

    // Test the result.
    vector<uint> spec;
    for (uint i  = 0; i < N / 2; i++) {
        spec.push_back(static_cast<uint>(i * fs + fmax));
        spec.push_back(static_cast<uint>((i + 1) * fs - fmax));
    }

    ASSERT_EQ(output.size(), N) << "Invalid size of the interpolated signal.";
    for (uint i = 0; i < N / 2; i++) {
        if (find(spec.begin(), spec.end(), i) != spec.end()) {
            ASSERT_NEAR(abs(output[i]), fs / 2, 1e-9) << "Spectral components must be at the multiples of fmax and fs - fmax.";
        }
        else { ASSERT_NEAR(abs(output[i]), 0, 1e-9) << "Noise must be close to 0."; }
    }
}

TEST_F(TestInterpolator, invalidInput) {
    double fs = -1000.0; // Sampling frequency must be positive.
    EXPECT_THROW(interpolator(60.0, 200.0, fs, {1}), invalid_argument);
}

File: tests/python/unit/test_unit_interpolator.py
import numpy as np
import pytest
from interpolator import Interpolator

# Define the test object.
@pytest.fixture
def interpolator():
    return Interpolator()

def test_interpolator_valid_output(mocker, interpolator):
    # Define the input.
    f_max = 400.00
    fs = 1000.00
    t = np.linspace(0, 1, int(fs), endpoint=False)
    input = np.cos(2 * np.pi * f_max * t)

    # Mock the HalfBand filter.
    halfband_result = [np.array([1])] * interpolator.N
    halfband_mock = mocker.Mock(side_effect=halfband_result)
    interpolator._Interpolator__halfband = halfband_mock

    # Compute the result.
    output = interpolator(120.0, f_max, fs, input)
    output = np.abs(np.fft.fft(output))
    N = 2 ** interpolator.N

    # Test the result.
    spec = []
    for i in range(N // 2):
        spec.append(int(i * fs + f_max))
        spec.append(int((i + 1) * fs - f_max))

    assert(output.size == N * input.size), "Invalid size of the interpolated signal."
    assert(np.allclose(output[spec], fs / 2, atol=1e-9)), "Spectral components must be at the multiples of f_max and fs - f_max."
    assert(np.allclose(np.delete(output[:output.size//2], spec), 0, atol=1e-9)), "Noise must be close to 0."

def test_interpolator__invalid_input(interpolator):
    with pytest.raises(ValueError, match="Interpolator.__call__: Sampling frequency must be positive."):
        interpolator(60.0, 200.0, -1000.0, np.array([1]))
